server/src/ws/routes.ts

```diff
@@ -1,50 +1,117 @@
-import { usersStorageClient } from '../storageClients.js'
 import { FastifyInstance } from 'fastify'
+import { WsMessageEvent } from 'shared/wsEvents.js'
+import { parseJwt } from '../utils/parseJwt.js'
+import { userClient } from '../userClient.js'
+import { User, UserId } from 'shared/userTypes.js'
+import { chatClient } from '../chatClient.js'
+import { chatGptClient } from '../chatGptClient.js'
+import { sleep } from 'shared/utils.js'
 
 export default async function wsHandler(server: FastifyInstance) {
-  // Auth is probably unnecessary, because to connect to the websocket you need to be logged in on the client anyway
-  // Uncomment this line to enable authentication
-  // server.addHook('preValidation', server.authenticate!) // TODO: make TS happy with this line
-
-  server.get<{ Querystring: { username: string } }>(
+  server.get<{ Querystring: { jwt: string } }>(
     '/',
     { websocket: true },
     async (connection, req) => {
       // A user join the chat
-      // const { id } = req.user as any
-      // const userAccounts = await usersStorageClient.load('allUsers')
-      // const user = userAccounts.find((account) => account.user.id === id)
-      // const username = user ? user.name : 'Unknown user'
-
-      broadcast({
-        sender: '__server',
-        message: `${'TODO'} joined`,
-      })
+      const { jwt } = req.query
+      const userId = getUserIdFromJwt(jwt)
+      const user = await getUserFromId(userId)
+      if (user) {
+        serverBroadcast(`${user.name} joined`)
+      }
 
       // A user leaving the chat
-      connection.socket.on('close', () => {
-        broadcast({
-          sender: '__server',
-          message: `${req.query.username} left`,
-        })
+      connection.socket.on('close', async () => {
+        const { jwt } = req.query
+        const userId = getUserIdFromJwt(jwt)
+        const user = await getUserFromId(userId)
+
+        if (user) {
+          serverBroadcast(`${user.name} left`)
+        }
       })
 
       // Broadcast incoming message
-      connection.socket.on('message', (message) => {
-        message = JSON.parse(message.toString())
+      connection.socket.on('message', async (msg) => {
+        msg = JSON.parse(msg.toString())
```

Comment:
This looks strange; what is msg initially? If it's already an object, you don't need to parse, and if it's already a string, you don't need toString

Reply:
Not sure, but I think it's raw bytes. That line is copied straight from the docs